contrib/grpc/template.py

```diff
@@ -0,0 +1,104 @@
+pkgname = "grpc"
+pkgver = "1.65.5"
+pkgrel = 0
+build_style = "cmake"
+configure_args = [
+    "-DgRPC_ABSL_PROVIDER=package",
+    "-DgRPC_BENCHMARK_PROVIDER=package",
+    "-DgRPC_CARES_PROVIDER=package",
+    "-DgRPC_PROTOBUF_PROVIDER=package",
+    "-DgRPC_RE2_PROVIDER=package",
+    "-DgRPC_SSL_PROVIDER=package",
+    "-DgRPC_ZLIB_PROVIDER=package",
+
+    "-DBUILD_SHARED_LIBS=ON",
+    "-DgRPC_BUILD_TESTS=ON",
+    "-DgRPC_INSTALL=ON",
+    "-DgRPC_DOWNLOAD_ARCHIVES=OFF",
+]
+hostmakedepends = [
+    "cmake",
+    "ninja",
+    "pkgconf"
+]
+makedepends = [
+    "abseil-cpp-devel",
+    "benchmark-devel",
+    "c-ares-devel",
+    "openssl-devel",
+    "protobuf-devel",
+    "re2-devel",
+    "zlib-ng-compat-devel",
+]
+checkdepends = [
+    "iproute2",
+    "procps",
+    "python-pyyaml"
+    "python-six",
+    "python-twisted",
+]
+pkgdesc = "High performance, open source universal RPC framework"
+maintainer = "sonata-chen <sonatachen39@gmail.com>"
+license = "Apache-2.0"
+url = "https://grpc.io"
+source = [
+    f"https://github.com/grpc/grpc/archive/refs/tags/v{pkgver}.tar.gz",
+    "https://github.com/google/googletest/archive/2dd1c131950043a8ad5ab0d2dda0e0970596586a.tar.gz",
+]
+source_paths = [".", "third_party/googletest"]
+sha256 = [
+    "bdab9cb52e3ee703d6c09d15716bcc10a150b2e9e414182e5b375d32050dba84",
+    "31bf78bd91b96dd5e24fab3bb1d7f3f7453ccbaceec9afb86d6e4816a15ab109",
+]
+
+
+def post_build(self):
+    special_tests = [
+        # RTNETLINK answers: Operation not permitted
+        "flaky_network_test",  # disabled
+
+        # it takes 58 mins to run
+        "grpc_tool_test",  # disabled
+
+        # need flags to run, so we run them manually
+        # tools/run_tests/generated/tests.json
+        "resolve_address_using_ares_resolver_posix_test",  # --resolver ares
+        "resolve_address_using_native_resolver_posix_test",  # --resolver native
+
+
+        # E0000 00:00:1724420309.061535      12 ssl_utils.cc:577] error loading file ./build/../../src/core/tsi/test_creds/ca.pem:
+        # INTERNAL: Failed to load file: ./build/../../src/core/tsi/test_creds/ca.pem due to error(fdopen): No such file or directory
+        #
+        # test/core/http/httpcli_test_util.cc
+        # we are going to run them with modified PATH environment variable
+        "httpcli_test",
+        "httpscli_test",
+    ]
+    with self.pushd("build"):
+        for t in special_tests:
+            self.mv(f"{t}", f"{t}_")
+
+
+def do_check(self):
+
+    script = R"""
+set -e
+python3 tools/run_tests/start_port_server.py
+
+build/resolve_address_using_ares_resolver_posix_test_ --resolver ares
+build/resolve_address_using_native_resolver_posix_test_ --resolver native
+
+PATH=./build:$PATH httpcli_test_
+PATH=./build:$PATH httpscli_test_
+
+find ./build -maxdepth 2 -name *_test -perm -111 -exec {} \;
+
+pkill -9 python3
```

Comment:
is there really no normal way to run the tests?

Reply:
These is a script `tools/run_tests/run_tests.py` in the grpc repo which can be used for testing during development. But that script always do a compilation for you.

Reply:
well i guess it's a start

instead of pkill -9 do

```
python3 ... &
pid=$!

...

kill $pid
```

Reply:
seems it also infinite loops for over 3 hours

```

E0000 00:00:1724480123.926900   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.927011   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.927143   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.927272   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.927407   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.927513   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.927643   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.927749   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.927878   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.927991   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.928118   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.928251   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.928387   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.928494   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
E0000 00:00:1724480123.928636   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
E0000 00:00:1724480123.928758   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
2024-08-24T06:15:23.9290675
```

i would guess that's a bug somewhere

Reply:
```
root@cbuild: /builddir/grpc-1.66.0$ python3 tools/run_tests/start_port_server.py 2> /dev/null &
root@cbuild: /builddir/grpc-1.66.0$ Port server started successfully

[1]   Done                    python3 tools/run_tests/start_port_server.py 2>/dev/null
root@cbuild: /builddir/grpc-1.66.0$
root@cbuild: /builddir/grpc-1.66.0$ echo $!
112
root@cbuild: /builddir/grpc-1.66.0$ ps -ejf
UID          PID    PPID    PGID     SID  C STIME TTY          TIME CMD
cbuild         1       0       0       0  0 06:04 ?        00:00:00 /usr/bin/bwrap --unshare-all --hostname cbuild --bind /home/billson/src/cports/bldroot / --bind /home/bil
cbuild         2       1       0       0  0 06:04 ?        00:00:00 sh /.cbuild_fakeroot.sh /usr/bin/sh -i
cbuild         4       1       4       4  0 06:04 ?        00:00:00 /usr/bin/faked
cbuild        11       2      11       0  0 06:04 ?        00:00:00 /usr/bin/sh -i
cbuild       113       1     113     113  0 07:39 ?        00:00:00 /usr/bin/python3 /builddir/grpc-1.66.0/tools/run_tests/python_utils/port_server.py -p 32766 -l /tmp/tmp2z
cbuild       115      11     115       0  0 07:40 ?        00:00:00 ps -ejf
root@cbuild: /builddir/grpc-1.66.0$
```

`tools/run_tests/start_port_server.py` forks a process(pid 113) and exits immediately. Maybe we
should try another method. Is there any way to detect that pid?

Reply:
probably doing nothing, actually?   
at the end of the .do invocation the stray chroot process are killed already, because of --die-with-parent

Reply:
Oh! That's really cool. i will remove that line then.

Reply:
> i would guess that's a bug somewhere

The many lines you mentioned come from tests that only take 3 minutes to run; it might not be a bug.
```
2024-08-24T06:15:22.1279412Z [==========] Running 1044 tests from 1 test suite.
2024-08-24T06:15:22.1280441Z [----------] Global test environment set-up.
2024-08-24T06:15:22.1281202Z [----------] 1044 tests from Http2SingleHopTest
2024-08-24T06:15:22.1281789Z [ RUN      ] Http2SingleHopTest.Encode0Decode0/Chttp2HttpProxy
2024-08-24T06:15:22.1470290Z WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
2024-08-24T06:15:22.1471600Z E0000 00:00:1724480122.146739   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
2024-08-24T06:15:22.1483664Z E0000 00:00:1724480122.148166   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
2024-08-24T06:15:22.1485615Z E0000 00:00:1724480122.148435   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
2024-08-24T06:15:22.1487091Z E0000 00:00:1724480122.148565   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
2024-08-24T06:15:22.1488701Z E0000 00:00:1724480122.148751   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
2024-08-24T06:15:22.1490205Z E0000 00:00:1724480122.148872   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
2024-08-24T06:15:22.1492354Z E0000 00:00:1724480122.149086   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
```
```
2024-08-24T06:18:16.4672956Z E0000 00:00:1724480296.466975   48038 cq_verifier.cc:368] Verify tag(101)-✅ for 10000ms
2024-08-24T06:18:16.4674287Z E0000 00:00:1724480296.467098   48038 cq_verifier.cc:368] Verify tag(102)-✅ tag(1)-✅ for 10000ms
2024-08-24T06:18:16.4677219Z E0000 00:00:1724480296.467563   48038 cq_verifier.cc:368] Verify tag(-1)-🤷 for 10000ms
2024-08-24T06:18:16.5685465Z [       OK ] Http2SingleHopTest.Encode4194304Decode4194304/Chttp2SslCredReloadTls12 (229 ms)
2024-08-24T06:18:16.5686851Z [----------] 1044 tests from Http2SingleHopTest (174440 ms total)
2024-08-24T06:18:16.5687426Z 
2024-08-24T06:18:16.5687733Z [----------] Global test environment tear-down
2024-08-24T06:18:16.5688729Z [==========] 1044 tests from 1 test suite ran. (174440 ms total)
2024-08-24T06:18:16.5689252Z [  PASSED  ] 1044 tests.
```

Reply:
> The many lines you mentioned come from tests that only take 3 minutes to run; it might not be a bug.

the entire thing was running for over three hours; either it's broken or we can't run the tests